server/lib/external-images.js

```diff
@@ -1,31 +1,43 @@
-const cheerio = require('cheerio');
 const Entities = require('html-entities').XmlEntities;
+const fetch = require('node-fetch');
 
 module.exports = function externalImages($) {
 	const entities = new Entities();
 
-	$('amp-img[src]').replaceWith((index, el) => {
-		const $el = cheerio(el).clone();
+	return Promise.all($('amp-img[src]').map((index, el) => {
+		const $el = $(el);
 		const matcher = /^https:\/\/h2.ft.com\/image\/v1\/images\/raw\/(.+)\?/;
 		const externalURI = $el.attr('src').match(matcher);
 
 		if(externalURI) {
 			const imageSrc = externalURI[1];
+			const imageSrcEncoded = encodeURIComponent(imageSrc);
 			// also unescape any html entites
-			const imageSrcEncoded = encodeURIComponent(entities.decode(imageSrc));
-			$el.attr('src', $el.attr('src').replace(imageSrc, imageSrcEncoded));
+			const fullSrcEncoded = entities.decode($el.attr('src').replace(imageSrc, imageSrcEncoded));
+			$el.attr('src', fullSrcEncoded);
 
-			// TODO: Properly set image dimensions.
-			// We *have* to specify a height and width for AMP to validate but we dont know
-			// what the height value will be. Width is set in the request to the image proxy
-			if(!$el.attr('height')) {
-				$el.attr('width', '700px');
-				$el.attr('height', '400px');
-			}
+			const metaUrl = fullSrcEncoded.replace('raw', 'metadata');
+			return fetch(metaUrl)
+				.then(response => response.json())
+				.then(
+					meta => Object.assign(meta, {ratio: meta.height / meta.width}),
+					(e) => (console.error(e), {width: 600, ratio: 4 / 7}) // discard error and use fallback dimensions
+				).then(meta => {
+					const width = Math.min(600, meta.width);
+					const height = width * meta.ratio;
+
+					$el.attr({
+						width: width,
+						height: height,
+					});
+
+					if(width < 600) {
+						// don't stretch narrow images to page width
+						$el.attr('layout', 'fixed');
+					}
+				});
 		}
 
 		return $el;
-	});
-
-	return $;
+	}).toArray()).then(() => $);
```

Comment:
Cheerio objects are array-like, not real arrays (they're akin to a `NodeList`), but `Promise.all` expects a real array or iterable.